zetaclient/bitcoin/txscript.go

```diff
@@ -0,0 +1,227 @@
+	}
+	if !IsPkScriptP2SH(script) {
+		return "", fmt.Errorf("invalid P2SH script: %s", scriptHex)
+	}
+	scriptHash := script[2:22]
+	return EncodeAddress(scriptHash, net.ScriptHashAddrID), nil
+}
+
+// DecodeScriptP2PKH decodes address from P2PKH script
+func DecodeScriptP2PKH(scriptHex string, net *chaincfg.Params) (string, error) {
+	script, err := hex.DecodeString(scriptHex)
+	if err != nil {
+		return "", errors.Wrapf(err, "error decoding script: %s", scriptHex)
+	}
+	if !IsPkScriptP2PKH(script) {
+		return "", fmt.Errorf("invalid P2PKH script: %s", scriptHex)
+	}
+	pubKeyHash := script[3:23]
+	return EncodeAddress(pubKeyHash, net.PubKeyHashAddrID), nil
+}
+
+// DecodeOpReturnMemo decodes memo from OP_RETURN script
+// returns (memo, found, error)
+func DecodeOpReturnMemo(scriptHex string, txid string) ([]byte, bool, error) {
+	if len(scriptHex) >= 4 && scriptHex[:2] == "6a" { // OP_RETURN
+		memoSize, err := strconv.ParseInt(scriptHex[2:4], 16, 32)
+		if err != nil {
+			return nil, false, errors.Wrapf(err, "error decoding memo size: %s", scriptHex)
+		}
+		if int(memoSize) != (len(scriptHex)-4)/2 {
+			return nil, false, fmt.Errorf("memo size mismatch: %d != %d", memoSize, (len(scriptHex)-4)/2)
+		}
+		memoBytes, err := hex.DecodeString(scriptHex[4:])
+		if err != nil {
+			return nil, false, errors.Wrapf(err, "error hex decoding memo: %s", scriptHex)
+		}
+		if bytes.Equal(memoBytes, []byte(common.DonationMessage)) {
+			return nil, false, fmt.Errorf("donation tx: %s", txid)
+		}
+		return memoBytes, true, nil
+	}
+	return nil, false, nil
+}
+
+// EncodeAddress returns a human-readable payment address given a ripemd160 hash
+// and netID which encodes the bitcoin network and address type.  It is used
+// in both pay-to-pubkey-hash (P2PKH) and pay-to-script-hash (P2SH) address
+// encoding.
+// Note: this function is a copy of the function in btcutil/address.go
+func EncodeAddress(hash160 []byte, netID byte) string {
+	// Format is 1 byte for a network and address class (i.e. P2PKH vs
+	// P2SH), 20 bytes for a RIPEMD160 hash, and 4 bytes of checksum.
+	return base58.CheckEncode(hash160[:ripemd160.Size], netID)
+}
+
+// DecodeTSSVout decodes receiver and amount from a given TSS vout
+func DecodeTSSVout(vout btcjson.Vout, receiverExpected string, chain common.Chain) (string, int64, error) {
+	// parse amount
+	amount, err := GetSatoshis(vout.Value)
+	if err != nil {
+		return "", 0, errors.Wrap(err, "error getting satoshis")
+	}
+	// get btc chain params
+	chainParams, err := common.GetBTCChainParams(chain.ChainId)
+	if err != nil {
+		return "", 0, errors.Wrapf(err, "error GetBTCChainParams for chain %d", chain.ChainId)
+	}
+	// decode cctx receiver address
+	addr, err := common.DecodeBtcAddress(receiverExpected, chain.ChainId)
+	if err != nil {
+		return "", 0, errors.Wrapf(err, "error decoding receiver %s", receiverExpected)
+	}
+	// parse receiver address from vout
+	var receiverVout string
+	switch addr.(type) {
```

Comment:
mentioned before but looks like this can be useful and can be added to common.

Reply:
same